### PR TITLE
Implemented gateway events GUILD_JOIN_REQUEST_UPDATE and GUILD_JOIN_REQUEST_DELETE.

### PR DESCRIPTION
**Pull request by bakirov-eldar**
Implemented gateway events GUILD_JOIN_REQUEST_UPDATE and GUILD_JOIN_REQUEST_DELETE.

These events are not officially documented by Discord, but they are actually sent by the Discord API and can be handled by bot gateway clients.

### CODE REVIEW

**Comment by KubaZ2** [comment]
Thanks for your contribution! Could you tell me when are the events sent?

**Comment by bakirov-eldar** [comment]
> Thanks for your contribution! Could you tell me when are the events sent?

GUILD_JOIN_REQUEST_UPDATE 
Fired when a user agrees to the member screening(rules)
GUILD_JOIN_REQUEST_DELETE
Fired when a user leaves without agreeing to the member screening(rules)

File: NetCord/Gateway/GatewayClient.cs [hunk 1]
```diff
@@ -81,6 +81,8 @@ public partial class GatewayClient : WebSocketClient, IEntity
     public event Func<Entitlement, ValueTask>? EntitlementUpdate;
     public event Func<Entitlement, ValueTask>? EntitlementDelete;
     public event Func<UnknownEventEventArgs, ValueTask>? UnknownEvent;
+    public event Func<GuildJoinRequestUpdateEventArgs, ValueTask>? GuildJoinRequestUpdate;
+    public event Func<GuildJoinRequestDeleteEventArgs, ValueTask>? GuildJoinRequestDelete;
```
Comment (KubaZ2) [inline NetCord/Gateway/GatewayClient.cs]:
```suggestion
    public event Func<GuildJoinRequestUpdateEventArgs, ValueTask>? GuildJoinRequestUpdate;
    public event Func<GuildJoinRequestDeleteEventArgs, ValueTask>? GuildJoinRequestDelete;
    public event Func<UnknownEventEventArgs, ValueTask>? UnknownEvent;
```

File: NetCord/Gateway/GuildJoinRequestStatus.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Text.Json.Serialization;
+
+namespace NetCord.Gateway;
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequestStatus.cs:3]:
```suggestion
namespace NetCord.Gateway;

```

File: NetCord/Gateway/GuildJoinRequestStatus.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+public enum GuildJoinRequestStatus
+{
+    [JsonPropertyName("STARTED")]
+    Started,
+    [JsonPropertyName("PENDING")]
+    Pending,
+    [JsonPropertyName("REJECTED")]
+    Rejected,
+    [JsonPropertyName("APPROVED")]
+    Approved
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequestStatus.cs]:
```suggestion
    Approved,
```

File: NetCord/Gateway/JsonModels/EventArgs/JsonGuildJoinRequestDeleteEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace NetCord.Gateway.JsonModels.EventArgs;
+
+public class JsonGuildJoinRequestDeleteEventArgs
+{
+    [JsonPropertyName("id")]
+    public ulong Id { get; set; }
+    [JsonPropertyName("user_id")]
+    public ulong UserId { get; set; }
+    [JsonPropertyName("guild_id")]
```
Comment (KubaZ2) [inline NetCord/Gateway/JsonModels/EventArgs/JsonGuildJoinRequestDeleteEventArgs.cs]:
```suggestion
    public ulong Id { get; set; }
    
    [JsonPropertyName("user_id")]
    public ulong UserId { get; set; }
    
    [JsonPropertyName("guild_id")]
```

File: NetCord/Gateway/JsonModels/JsonGuildJoinRequest.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    public GuildJoinRequestStatus ApplicationStatus { get; set; }
+
+    [JsonPropertyName("created_at")]
+    public DateTimeOffset CreatedAt { get; set; }
+
+    [JsonPropertyName("guild_id")]
+    public ulong GuildId { get; set; }
+
+    [JsonPropertyName("last_seen")]
+    public DateTimeOffset LastSeen { get; set; }
```
Comment (KubaZ2) [inline NetCord/Gateway/JsonModels/JsonGuildJoinRequest.cs]:
```suggestion
    public DateTimeOffset LastSeenAt { get; set; }
```

File: NetCord/Gateway/JsonModels/JsonGuildJoinRequest.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Text.Json.Serialization;
+
+using NetCord.JsonModels;
+
+namespace NetCord.Gateway.JsonModels;
```
Comment (KubaZ2) [inline NetCord/Gateway/JsonModels/JsonGuildJoinRequest.cs:5]:
```suggestion
namespace NetCord.Gateway.JsonModels;

```

File: NetCord/Gateway/JsonModels/JsonVerificationField.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Text.Json.Serialization;
+
+namespace NetCord.Gateway.JsonModels;
```
Comment (KubaZ2) [inline NetCord/Gateway/JsonModels/JsonVerificationField.cs]:
```suggestion
namespace NetCord.Gateway.JsonModels;

```

File: NetCord/Gateway/VerificationField.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+public class VerificationField(JsonModels.JsonVerificationField jsonModel) : IJsonModel<JsonModels.JsonVerificationField>
+{
+    private readonly JsonModels.JsonVerificationField _jsonModel = jsonModel;
+    JsonModels.JsonVerificationField IJsonModel<JsonModels.JsonVerificationField>.JsonModel => _jsonModel;
+    public VerificationFieldType FieldType => _jsonModel.FieldType;
+    public string Label => _jsonModel.Label;
+    public bool Required => _jsonModel.Required;
+    public bool IsResponse => _jsonModel.Response;
+    public IReadOnlyList<string> Values => _jsonModel.Values;
```
Comment (KubaZ2) [inline NetCord/Gateway/VerificationField.cs]:
```suggestion
    JsonModels.JsonVerificationField IJsonModel<JsonModels.JsonVerificationField>.JsonModel => jsonModel;
    
    public VerificationFieldType FieldType => jsonModel.FieldType;
    public string Label => jsonModel.Label;
    public bool Required => jsonModel.Required;
    public bool IsResponse => jsonModel.Response;
    public IReadOnlyList<string> Values => jsonModel.Values;
```

File: NetCord/Gateway/VerificationFieldType.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Text.Json.Serialization;
+
+namespace NetCord.Gateway;
```
Comment (KubaZ2) [inline NetCord/Gateway/VerificationFieldType.cs]:
```suggestion
namespace NetCord.Gateway;

```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+﻿using NetCord.Rest;
+
+namespace NetCord.Gateway;
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs:3]:
```suggestion
namespace NetCord.Gateway;

```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+public class GuildJoinRequest : IJsonModel<JsonModels.JsonGuildJoinRequest>
+{
+    private readonly JsonModels.JsonGuildJoinRequest _jsonModel;
+    JsonModels.JsonGuildJoinRequest IJsonModel<JsonModels.JsonGuildJoinRequest>.JsonModel => _jsonModel;
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs]:
```suggestion
    JsonModels.JsonGuildJoinRequest IJsonModel<JsonModels.JsonGuildJoinRequest>.JsonModel => _jsonModel;
    
```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    public User User { get; }
+    public IReadOnlyList<VerificationField> FormResponses { get; }
+    public User ActionedByUser { get; }
+    public ulong ActionedAt => _jsonModel.ActionedAt;
+    public GuildJoinRequest(JsonModels.JsonGuildJoinRequest jsonModel, RestClient client)
+    {
+        _jsonModel = jsonModel;
+        User = new User(jsonModel.User, client);
+        ActionedByUser = new User(jsonModel.ActionedByUser, client);
+        FormResponses = _jsonModel.FormResponses.Select(e => new VerificationField(e)).ToList();
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs]:
```suggestion
        FormResponses = _jsonModel.FormResponses.Select(e => new VerificationField(e)).ToArray();
```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    public GuildJoinRequestStatus ApplicationStatus => _jsonModel.ApplicationStatus;
+    public DateTimeOffset? CreatedAt => _jsonModel.CreatedAt;
+    public ulong GuildId => _jsonModel.GuildId;
+    public DateTimeOffset LastSeen => _jsonModel.LastSeen;
+    public string? RejectionReason => _jsonModel.RejectionReason;
+    public ulong UserId => _jsonModel.UserId;
+    public User User { get; }
+    public IReadOnlyList<VerificationField> FormResponses { get; }
+    public User ActionedByUser { get; }
+    public ulong ActionedAt => _jsonModel.ActionedAt;
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs]:
```suggestion
    public ulong ActionedAt => _jsonModel.ActionedAt;
    
```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    public GuildJoinRequest(JsonModels.JsonGuildJoinRequest jsonModel, RestClient client)
+    {
+        _jsonModel = jsonModel;
+        User = new User(jsonModel.User, client);
+        ActionedByUser = new User(jsonModel.ActionedByUser, client);
+        FormResponses = _jsonModel.FormResponses.Select(e => new VerificationField(e)).ToList();
+    }
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs]: I would use a primary constructor here

File: NetCord/Serialization.cs [hunk 15]
```diff
@@ -223,6 +223,12 @@ namespace NetCord;
 [JsonSerializable(typeof(GuildUsersSearchTimestamp))]
 [JsonSerializable(typeof(JsonGuildUserSearchResult))]
 [JsonSerializable(typeof(GuildUsersSearchPaginationProperties))]
+[JsonSerializable(typeof(IReadOnlyList<string>))]
+[JsonSerializable(typeof(JsonVerificationField))]
+[JsonSerializable(typeof(IReadOnlyList<JsonVerificationField>))]
+[JsonSerializable(typeof(JsonGuildJoinRequest))]
```
Comment (KubaZ2) [inline NetCord/Serialization.cs]:
```suggestion
```

File: NetCord/Gateway/VerificationFieldType.cs [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    [JsonPropertyName("TEXT_INPUT")]
+    TextInput,
+    [JsonPropertyName("PARAGRAPH")]
+    Paragraph,
+    [JsonPropertyName("MULTIPLE_CHOICE")]
+    MultipleChoise,
+    [JsonPropertyName("VERIFICATION")]
+    Verification,
+    [JsonPropertyName("FILE_UPLOAD")]
+    FileUpload
```
Comment (KubaZ2) [inline NetCord/Gateway/VerificationFieldType.cs]:
```suggestion
    FileUpload,
```

File: NetCord/Gateway/VerificationFieldType.cs [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+public enum VerificationFieldType
+{
+    [JsonPropertyName("TERMS")]
+    Terms,
+    [JsonPropertyName("TEXT_INPUT")]
+    TextInput,
+    [JsonPropertyName("PARAGRAPH")]
+    Paragraph,
+    [JsonPropertyName("MULTIPLE_CHOICE")]
+    MultipleChoise,
```
Comment (KubaZ2) [inline NetCord/Gateway/VerificationFieldType.cs]:
```suggestion
    MultipleChoice,
```

File: NetCord/Gateway/GuildJoinRequest.cs [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace NetCord.Gateway;
+public class GuildJoinRequest : IJsonModel<JsonModels.JsonGuildJoinRequest>
+{
+    private readonly JsonModels.JsonGuildJoinRequest _jsonModel;
+    JsonModels.JsonGuildJoinRequest IJsonModel<JsonModels.JsonGuildJoinRequest>.JsonModel => _jsonModel;
+    public GuildJoinRequestStatus ApplicationStatus => _jsonModel.ApplicationStatus;
+    public DateTimeOffset? CreatedAt => _jsonModel.CreatedAt;
+    public ulong GuildId => _jsonModel.GuildId;
+    public DateTimeOffset LastSeen => _jsonModel.LastSeen;
```
Comment (KubaZ2) [inline NetCord/Gateway/GuildJoinRequest.cs]:
```suggestion
    public DateTimeOffset LastSeenAt => _jsonModel.LastSeenAt;
```

**Comment by KubaZ2** [comment]
@bakirov-eldar, is there any reason you used `VerificationField` name? Based on the API, looks like `FormResponse` would be more appropriate.

EDIT: I renamed it to `GuildJoinRequestFormResponse` in https://github.com/KubaZ2/NetCord/pull/15/commits/a934db4f14e8532811b1fa596497fb8116829d98

**Pull request merged**